src/services/upload.ts

```diff
@@ -0,0 +1,125 @@
+import { storage } from "~/firebase/init";
+import {
+  ref,
+  uploadBytes,
+  getDownloadURL,
+  FirebaseStorage,
+  listAll,
+  StorageReference,
+  uploadBytesResumable,
+  deleteObject,
+} from "firebase/storage";
+import { Dispatch } from "react";
+
+type UploadFilePath = {
+  userId: string;
+  file: File;
+};
+type UpdateFile = {
+  filePath: string;
+  file: File;
+};
+
+class UploadService {
+  public readonly rootFolder: string = "jazzapp/music/";
+
+  public async uploadFile(
+    fPath: UploadFilePath,
+    setProgress?: Dispatch<{ percents: number; operation: string }>,
+  ): Promise<{ status: boolean; url?: string }> {
+    if (!fPath.file) return Promise.resolve({ status: false });
+
+    const subMusicRef = `/${fPath.userId}/${fPath.file.name}`;
+    const fullMusicRef = ref(storage, `${this.rootFolder}${subMusicRef}`);
+    const uploadTask = uploadBytesResumable(fullMusicRef, fPath.file);
+
+    uploadTask.on(
+      "state_changed",
+      (snapshot) => {
+        const progress = Math.round(
+          (snapshot.bytesTransferred / snapshot.totalBytes) * 100,
+        );
+        if (setProgress) {
+          setProgress({ percents: progress, operation: "Upload" });
+        }
```

Comment:
Try to use this technique:
```tsx
setProgress?.( ... )
```

[Read more](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining#optional_chaining_with_function_calls)